loadtest/loadtest_client.go

```diff
@@ -108,194 +91,58 @@ func (c *LoadTestClient) Close() {
-	senders := []func(){}
-	if config.RunOracle && i%2 == 0 {
-		for _, valKey := range valKeys {
-			// generate oracle tx
-			msg := generateOracleMessage(valKey)
-			txBuilder := TestConfig.TxConfig.NewTxBuilder()
-			_ = txBuilder.SetMsgs(msg)
-			seqDelta := uint64(i / 2)
-			mode := typestx.BroadcastMode_BROADCAST_MODE_SYNC
-			sender := SendTx(valKey, &txBuilder, mode, seqDelta, false, *c, 30000, 100000)
-			waitGroup.Add(1)
-			senders = append(senders, func() {
-				defer waitGroup.Done()
-				sender()
-			})
 		}
 	}
-	return senders
 }
 
-func (c *LoadTestClient) SendTxs(workgroups []*sync.WaitGroup, sendersList [][]func()) {
-	defer close(c.TxResponseChan)
+func (c *LoadTestClient) SendTxs(txQueue <-chan []byte, consumerId int, wg *sync.WaitGroup, done <-chan struct{}, sentCount *int64) {
+	defer wg.Done()
 
-	lastHeight := getLastHeight(c.LoadTestConfig.BlockchainEndpoint)
-	for i := 0; i < int(c.LoadTestConfig.Rounds); i++ {
-		newHeight := getLastHeight(c.LoadTestConfig.BlockchainEndpoint)
-		for newHeight == lastHeight {
-			time.Sleep(10 * time.Millisecond)
-			newHeight = getLastHeight(c.LoadTestConfig.BlockchainEndpoint)
-		}
-		fmt.Printf("Sending %d-th block\n", i)
-		senders := sendersList[i]
-		wg := workgroups[i]
-		for _, sender := range senders {
-			go sender()
-		}
-		wg.Wait()
-		lastHeight = newHeight
-	}
-}
+	for {
 
-func (c *LoadTestClient) GatherTxHashes() {
-	for txHash := range c.TxResponseChan {
-		c.TxHashList = append(c.TxHashList, *txHash)
-	}
-	fmt.Printf("Transactions Sent=%d\n", len(c.TxHashList))
-}
-
-func (c *LoadTestClient) ValidateTxs() {
-	defer c.Close()
-	numTxs := len(c.TxHashList)
-	resultChan := make(chan *types.TxResponse, numTxs)
-	var waitGroup sync.WaitGroup
-
-	if numTxs == 0 {
-		return
-	}
-
-	for _, txHash := range c.TxHashList {
-		waitGroup.Add(1)
-		go func(txHash string) {
-			defer waitGroup.Done()
-			resultChan <- c.GetTxResponse(txHash)
-		}(txHash)
-	}
-
-	go func() {
-		waitGroup.Wait()
-		close(resultChan)
-	}()
-
-	fmt.Printf("Validating %d Transactions... \n", len(c.TxHashList))
-	waitGroup.Wait()
-
-	notCommittedTxs := 0
-	responseCodeMap := map[int]int{}
-	responseStringMap := map[string]int{}
-	for result := range resultChan {
-		// If the result is nil then that means the transaction was not committed
-		if result == nil {
-			notCommittedTxs++
-			continue
-		}
-		code := result.Code
-		codeString := "ok"
-		if code != 0 {
-			codespace := result.Codespace
-			err := sdkerrors.ABCIError(codespace, code, fmt.Sprintf("Error code=%d ", code))
-			codeString = err.Error()
+		select {
+		case <-done:
+			fmt.Printf("Stopping consumer %d\n", consumerId)
+			return
+		case tx, ok := <-txQueue:
+			if !ok {
+				fmt.Printf("Stopping consumer %d\n", consumerId)
+			}
+			SendTx(tx, typestx.BroadcastMode_BROADCAST_MODE_SYNC, false, *c, sentCount)
```

Comment:
we likely want this to be async, right? otherwise there would be a bottleneck for sending txs since theyll need to sequentially pass checkTx to enter mempool?

Reply:
We want to do sync here and wait until the response back, otherwise you have to query the tx and figure out when it is completed. We can put each tx into a separate goroutine though